physics-informed-neural-networks-for-steady-cavity-flow/README.md

```diff
@@ -0,0 +1,274 @@
+[psiTest, pTest, uTest, vTest] = dlfeval(@calculateStreamfunctionPressureAndVelocity, net, xTest, yTest);
+
+% Return predictions to grid and plot.
+ut = unflattenAndExtract(uTest, numTestSamples);
+vt = unflattenAndExtract(vTest, numTestSamples);
+pt = unflattenAndExtract(pTest, numTestSamples);
+psit = unflattenAndExtract(psiTest, numTestSamples);
+
+figure; 
+subplot(2,2,1)
+contourf(xt, yt, psit)
+colorbar
+axis equal
+title('psi')
+
+subplot(2,2,2)
+contourf(xt, yt, pt)
+colorbar
+axis equal
+title('p')
+
+subplot(2,2,3)
+contourf(xt, yt, ut)
+colorbar
+axis equal
+title('u')
+
+subplot(2,2,4)
+contourf(xt, yt, vt)
+colorbar
+axis equal
+title('v')
+```
+
+![figure_0.png](README_media/figure_0.png)
+# Loss function and helper functions
+```matlab
+function [loss, grads, lossEqnX, lossEqnY, lossBC] = pinnsLossFunction(net, xyEquation, xyBoundary, zeroVector, uvBoundary, Re)
+
+% Get model outputs at interior points.
+xeq = xyEquation(:, 1);
+yeq = xyEquation(:, 2);
+[psi, p] = forward(net, xeq, yeq);
+
+% Compute gradients.
+psisum = sum(psi,1);
+u = dlgradient(psisum, yeq, EnableHigherDerivatives=true);
+v = -1.*dlgradient(psisum, xeq, EnableHigherDerivatives=true);
+
+usum = sum(u,1);
+ux = dlgradient(usum, xeq, EnableHigherDerivatives=true);
+uy = dlgradient(usum, yeq, EnableHigherDerivatives=true);
+uxx = dlgradient(sum(ux,1), xeq, EnableHigherDerivatives=true);
+uyy = dlgradient(sum(uy,1), yeq, EnableHigherDerivatives=true);
+
+vsum = sum(v,1);
+vx = dlgradient(vsum, xeq, EnableHigherDerivatives=true);
+vy = dlgradient(vsum, yeq, EnableHigherDerivatives=true);
+vxx = dlgradient(sum(vx,1), xeq, EnableHigherDerivatives=true);
+vyy = dlgradient(sum(vy,1), yeq, EnableHigherDerivatives=true);
+
+psum = sum(p,1);
+px = dlgradient(psum, xeq, EnableHigherDerivatives=true);
+py = dlgradient(psum, yeq, EnableHigherDerivatives=true);
+
+% Momentum equations.
+lx = u.*ux + v.*uy + px - (1/Re).*(uxx + uyy);
+ly = u.*vx + v.*vy + py - (1/Re).*(vxx + vyy);
+
+% Combine for equation loss.
+lossEqnX = logCoshLoss(lx, zeroVector);
+lossEqnY = logCoshLoss(ly, zeroVector);
+
+% Get model outputs at boundary points.
+xbd = xyBoundary(:, 1);
+ybd = xyBoundary(:, 2);
+psibd = forward(net, xbd, ybd);
+
+psibdsum = sum(psibd,1);
+ubd = dlgradient(psibdsum, ybd, EnableHigherDerivatives=true);
+vbd = -1.*dlgradient(psibdsum, xbd, EnableHigherDerivatives=true);
+
+uvbd = cat(2, ubd, vbd);
+lossBC = logCoshLoss(uvbd, uvBoundary);
+
+% Total loss and model gradients
+loss = lossEqnX + lossEqnY + lossBC;
+grads = dlgradient(loss, net.Learnables);
+end
+
+function loss = logCoshLoss(y, t)
+% log-cosh loss function
+e = y - t;
+loss = mean( log(cosh(e)), 'all' );
+end
+
+function [psi, p, u, v] = calculateStreamfunctionPressureAndVelocity(net, x, y)
+% Compute the streamfunction psi, pressure p and velocity (u,v) given
+% input positions (x,y).
+[psi, p] = forward(net, x, y);
+psisum = sum(psi,1);
+u = dlgradient(psisum, y);
+v = -1.*dlgradient(psisum, x);
```

Comment:
It could be worth making this one `dlgradient` call, something like
```
[u,v] = dlgradient(psisum,y,x);
v = -v;
```
I had some feeling this was faster, although I'm not completely confident. I'm sure I found it faster when I did something like batching the interior points and boundary points together and performing one `dlgradient`, then splitting the output back into interior and boundary gradients. There could be a few places to try that out in the loss above.